flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/BigQuerySchemaProvider.java

```diff
@@ -0,0 +1,379 @@
+import com.google.cloud.flink.bigquery.common.config.BigQueryConnectOptions;
+import com.google.cloud.flink.bigquery.common.exceptions.BigQueryConnectorException;
+import com.google.cloud.flink.bigquery.common.utils.SchemaTransform;
+import com.google.cloud.flink.bigquery.services.BigQueryServices.QueryDataClient;
+import com.google.cloud.flink.bigquery.services.BigQueryServicesFactory;
+import com.google.cloud.flink.bigquery.services.BigQueryUtils;
+import com.google.protobuf.DescriptorProtos.DescriptorProto;
+import com.google.protobuf.DescriptorProtos.FieldDescriptorProto;
+import com.google.protobuf.DescriptorProtos.FileDescriptorProto;
+import com.google.protobuf.Descriptors.Descriptor;
+import com.google.protobuf.Descriptors.DescriptorValidationException;
+import com.google.protobuf.Descriptors.FileDescriptor;
+import org.apache.avro.LogicalType;
+import org.apache.avro.LogicalTypes;
+import org.apache.avro.Schema;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+
+import javax.annotation.Nullable;
+
+import java.io.Serializable;
+import java.util.EnumMap;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ * Class to obtain Descriptor ({@link Descriptor}) required for serialization of Generic Records.
+ * This class has essential methods to obtain bigQuery Table Schema ({@link TableSchema}) from the
+ * provided Table Connection Options ({@link BigQueryConnectOptions}). It is also responsible for
+ * converting this Table Schema to Avro Schema ({@link Schema}) which is used to further convert to
+ * DescriptorProto({@link DescriptorProto}). The obtained Descriptor Proto is required for the
+ * formation of the Descriptors which is essential for Proto Rows formation.
+ */
+public class BigQuerySchemaProvider implements Serializable {
+
+    private Schema avroSchema;
+    private DescriptorProto descriptorProto;
+
+    private static final Map<Schema.Type, FieldDescriptorProto.Type> AVRO_TYPES_TO_PROTO;
+    private static final Map<String, FieldDescriptorProto.Type> LOGICAL_AVRO_TYPES_TO_PROTO;
+
+    public DescriptorProto getDescriptorProto() {
+        return descriptorProto;
+    }
+
+    public Descriptor getDescriptor() {
+        Preconditions.checkNotNull(
+                getDescriptorProto(),
+                "DescriptorProto not initialized before obtaining Descriptor!");
+        try {
+            return getDescriptorFromDescriptorProto(descriptorProto);
+        } catch (DescriptorValidationException e) {
+            throw new BigQueryConnectorException(
+                    "Error obtaining Descriptor from Descriptor Proto", e.getCause());
+        }
+    }
+
+    public Schema getSchema() {
+        return this.avroSchema;
+    }
+
+    public BigQuerySchemaProvider(BigQueryConnectOptions connectOptions) {
+        this.getBigQuerySchemaProvider(connectOptions);
+    }
```

Comment:
This wrapping is not needed. Move the contents of `getBigQuerySchemaProvider(connectOptions)` to this method.